content/2016-04-18-aliasing-improved.rst

```diff
@@ -0,0 +1,473 @@
+           textual representation of the alias set of the ``ast.BoolOp`` nodes,
+           which turns out to be ``['a', 'b']``.
+
+Unary Operator Expression
+-------------------------
+
+Resulting node does not alias to anything
+
+.. code:: python
+
+    >>> module = ast.parse('def foo(a): return -a')
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.UnaryOp)
+    (- a) => ['<unbound-value>']
+
+As stated previously, values not bound to an identifier are only represented as ``<unbound-value>``.
+
+If Expression
+-------------
+
+Resulting node alias to either branch
+
+.. code:: python
+
+    >>> module = ast.parse('def foo(a, b, c): return a if c else b')
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.IfExp)
+    (a if c else b) => ['a', 'b']
+
+Dict Expression
+---------------
+
+A dict is abstracted as an unordered container of its values
+
+.. code:: python
+
+    >>> module = ast.parse('def foo(a, b): return {0: a, 1: b}')
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.Dict)
+    {0: a, 1: b} => ['|a|', '|b|']
+
+where the ``|id|`` notation means something that may contain ``id``.
+
+Set Expression
+--------------
+
+
+A set is abstracted as an unordered container of its elements
+
+.. code:: python
+
+    >>> module = ast.parse('def foo(a, b): return {a, b}')
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.Set)
+    {a, b} => ['|a|', '|b|']
+
+List Expression
+---------------
+
+
+A list is abstracted as an ordered container of its values
+
+.. code:: python
+
+    >>> module = ast.parse('def foo(a, b): return [a, b]')
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.List)
+    [a, b] => ['|[0]=a|', '|[1]=b|']
+
+where the ``|[i]=id|`` notation means something that
+may contain ``id`` at index ``i``.
+
+
+Call Expression
+---------------
+
+Resulting node alias to the return_alias of called function,
+if the function is already known by Pythran (i.e. it's an Intrinsic)
+or if Pythran already computed it's ``return_alias`` behavior.
+
+.. code:: python
+
+    >>> fun0 = 'def f(a): return a\n'
+    >>> fun1 = 'def foo(b): c = f(b)\n'
+    >>> module = ast.parse(fun0 + fun1)
+
+The ``f`` function create aliasing between
+the returned value and its first argument.
+
+.. code:: python
+
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.Call)
+    f(b) => ['b']
+
+This also works with intrinsics, e.g ``dict.setdefault`` which
+may create alias between its third argument and the return value
+
+.. code:: python
+
+    >>> fun = 'def foo(a, d): __builtin__.dict.setdefault(d, 0, a)\n'
+    >>> module = ast.parse(fun)
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.Call)
+    __builtin__.dict.setdefault(d, 0, a) => ['<unbound-value>', 'a']
```

Comment:
Yeh, we should if we supported indexed aliasing for dict, which is not
the case :-/